### DISCUSSION TITLE
Network Packet Size

### DISCUSSION THREAD

**Discussion by oldiaz507** [sqlalchemy/sqlalchemy#10708] [Usage Questions] [answered]
Hi, I’m hoping somebody here can help me. 
I have a working program that connects to SAP ASE (formerly Sybase ASE) and extracts data. It works as expected but it's using the default network packet size.
In order to speed up the process I need to use a larger packet size and have been looking for a week and can't find any documentation.

Is it possible to specify a packet size?

### Setup
We have a RHEL 8 with Python 3.9 and the following packages:
```

Package           Version
----------------- ------------
cycler            0.12.1
fonttools         4.45.0
greenlet          1.1.3.post0
joblib            1.3.2
kiwisolver        1.3.1
matplotlib        3.5.3
numpy             1.22.4
odbcinst          1.0.1
packaging         23.2
pandas            2.1.3
Pillow            8.3.2
pip               23.3.1
pyarrow           14.0.1
pyodbc            5.0.1
pyparsing         3.1.1
python-dateutil   2.8.2
pytz              2023.3.post1
scikit-learn      1.0.1
scipy             1.7.1
seaborn           0.12.2
setuptools        50.3.2
six               1.16.0
SQLAlchemy        2.0.23
sqlalchemy-sqlany 1.0.3
sqlalchemy-sybase 2.0.0
sqlanydb          1.0.6.1
threadpoolctl     3.2.0
typing_extensions 4.8.0
tzdata            2023.3
```

Thanks in advance.

**Comment by zzzeek** [comment]
this would be a factor of the Sybase DBAPI you're using and maybe even OS level settings.    CCing @gordthompson who maintains https://github.com/gordthompson/sqlalchemy-sybase

**Comment by gordthompson** [comment, marked as answer]
The pyodbc [documentation](https://github.com/mkleehammer/pyodbc/wiki/The-pyodbc-Module) mentions a way to change the packet size using the `attrs_before=` parameter in the `pyodbc.connect()` method.

I am not able to try that with SAP ASE (and I am no longer able to actively support the sqlalchemy-sybase dialect as I announced [here](https://github.com/gordthompson/sqlalchemy-sybase/issues/11) last March), but I did try it with mssql using

```python
def get_connection():
    SQL_ATTR_PACKET_SIZE = 112
    return pyodbc.connect(
        "Driver=ODBC Driver 17 for SQL Server;"
        "Server=192.168.0.199;"
        "UID=scott;PWD=tiger^5HHH;"
        "Database=test;",
        attrs_before={SQL_ATTR_PACKET_SIZE: 1024 * 8},
    )


engine = sa.create_engine("mssql+pyodbc://", creator=get_connection)
```

Increasing the `SQL_ATTR_PACKET_SIZE` on the client side did not help — in fact it drastically degraded performance. Still, you might want to give it a try and see if SAP/Sybase likes it better.

**Reply by oldiaz507** [reply]
Thank you Gord!

This worked perfectly with ASE for downloading many big tables to disk.

```
def get_connection():
	SQL_ATTR_PACKET_SIZE = 112
	return pyodbc.connect("Driver={Adaptive Server Enterprise};Server=Servidor;Port=7500;UID=Usuario;PWD=Clave;",
        attrs_before={SQL_ATTR_PACKET_SIZE: 16384},
    )

engine = sa.create_engine("sybase+pyodbc://", creator=get_connection)
```
